File: LeetCode/Backtracking/JavaScript/wordSearch.js
// Given an m x n grid of characters board and a string word, return true if word exists in the grid.

// The word can be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once.

// Example 1:

// Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCCED"
// Output: true
// Example 2:

// Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "SEE"
// Output: true
// Example 3:

// Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCB"
// Output: false

// Constraints:

// m == board.length
// n = board[i].length
// 1 <= m, n <= 6
// 1 <= word.length <= 15
// board and word consists of only lowercase and uppercase English letters.

// Follow up: Could you use search pruning to make your solution faster with a larger board?

// TC: O(n*m*4^(length of word))
// SC: O(n*m*4^(length of word))

const Result = { PASS: "\x1b[92mPASS\x1b[0m", FAIL: "\x1b[91mFAIL\x1b[0m" };

class WordSearchRecord {
  constructor(board, word, expected) {
    this.board = board;
    this.word = word;
    this.expected = expected;
  }
}

class Solution {
  /**
   * @param {character[][]} board
   * @param {string} word
   * @return {boolean}
   */
  exist(board, word) {
    let rows = board.length;
    let cols = board[0].length;
    let path = new Set();

    // DFS on 4 dirs and backtrack to brute force all cells. Pass in row, col and index of the char of the word
    // Base case when the index of the char of the word we're looking is equal to the length of the word. Return true
    // Other cases: Out of bounds for row and col, row and col are in our visited set, word[index] is not in the board[row][col], return false
    const backtrack = (r, c, i) => {
      if (i === word.length) return true;
      if (
        r < 0 ||
        c < 0 ||
        r >= rows ||
        c >= cols ||
        word[i] !== board[r][c] ||
        path.has(`${r}${c}`)
      )
        return false;

      path.add(`${r}${c}`);
      // DFS on 4 dirs and backtrack to brute force all cells
      let res =
        backtrack(r + 1, c, i + 1) ||
        backtrack(r - 1, c, i + 1) ||
        backtrack(r, c + 1, i + 1) ||
        backtrack(r, c - 1, i + 1);
      path.delete(`${r}${c}`);
      return res;
    };

    // Brute force and try to visit all cells
    for (let r = 0; r < rows; r += 1) {
      for (let c = 0; c < cols; c += 1) {
        if (backtrack(r, c, 0)) return true;
      }
    }

    return false;
  }
}

function testSolution(record) {
  const solution = new Solution();
  const result = solution.exist(record.board, record.word);
  const pass = result === record.expected;

  console.log(`Input: board = ${JSON.stringify(record.board)}`);
  console.log(`       word = ${JSON.stringify(record.word)}`);
  console.log(`Expected: ${record.expected}`);
  console.log(`Result: ${result}`);
  console.log(pass ? Result.PASS : Result.FAIL);
}

const records = [
  new WordSearchRecord(
    [
      ["A", "B", "C", "E"],
      ["S", "F", "C", "S"],
      ["A", "D", "E", "E"],
    ],
    "ABCCED",
    true
  ),
  new WordSearchRecord(
    [
      ["A", "B", "C", "E"],
      ["S", "F", "C", "S"],
      ["A", "D", "E", "E"],
    ],
    "SEE",
    true
  ),
  new WordSearchRecord(
    [
      ["A", "B", "C", "E"],
      ["S", "F", "C", "S"],
      ["A", "D", "E", "E"],
    ],
    "ABCB",
    false
  ),
  new WordSearchRecord([["A"]], "A", true),
  new WordSearchRecord([["A"]], "AA", false),
];

records.forEach((record, index) => {
  console.log(`# Test case ${index + 1}`);
  testSolution(record);
  console.log("----------------------------------------");
});

File: LeetCode/Backtracking/python/wordSearch.py
from dataclasses import dataclass
from enum import Enum
from typing import List

class Result(Enum):
    PASS = "\033[92mPASS\033[00m"
    FAIL = "\033[91mFAIL\033[00m"

@dataclass
class wordSearchRecord:
    board: List[List[str]]
    word: str
    expected: bool

class Solution:
    def __init__(self):
        self.numRows = 0
        self.numCols = 0
        self.path = set()

    def exist(self, board: List[List[str]], word: str) -> bool:
        self.numRows = len(board)
        self.numCols = len(board[0])

        for r in range(self.numRows):
            for c in range(self.numCols):
                if self.backtrack(r, c, 0, board, word):
                    return True
                
        return False
    
    def backtrack(self, r: int, c: int, i: int, board: List[List[str]],  word: str) -> bool:
        if i == len(word):
            return True
        
        # Check if out of bound
        if r < 0 or c < 0 or r >= self.numRows or c >= self.numCols:
            return False
        
        # Check if word[i] 
        if word[i] != board[r][c]:
            return False
        
        # Check if we've seen this before
        if (r,c) in self.path:
            return False
        
        # Add the tuple r,c to the path
        self.path.add((r,c))

        # DFS on 4 dirs and backtrack to bruteforce all cells
        res: bool = self.backtrack(r + 1, c, i + 1, board, word) or \
                    self.backtrack(r - 1, c, i + 1, board, word) or \
                    self.backtrack(r, c + 1, i + 1, board, word) or \
                    self.backtrack(r, c - 1, i + 1, board, word)
        
        # Remove the tuple r,c from the path
        self.path.remove((r,c))

        return res

    @staticmethod
    def testSolution(record: wordSearchRecord) -> None:
        print(f"input:\tboard {record.board}")
        print(f"\tword: {record.word}")
        print(f"expected: {record.expected}")

        res: bool = Solution().exist(record.board, record.word)

        print(f"result: {res}")
        print(Result.PASS.value if res == record.expected else Result.FAIL.value)

if __name__ == "__main__":
    records: List[wordSearchRecord] = [
        wordSearchRecord([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCCED", True),
        wordSearchRecord([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "SEE", True),
        wordSearchRecord([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCB", False),
    ]

    for i, record in enumerate(records):
        print(f"# Test case {i + 1}")
        Solution.testSolution(record)
        print("-" * 50)